### PR TITLE
feat: Moar scalable

### PR DESCRIPTION
**Pull request by kgilpin**
Also, a bunch of fixes. See individual commits to best understand what's in here.

This code was written primarily as I was refactoring the Solidus support. It's a large project which demanded some better performance and behavior from our extension. It also uses a monorepo structure, which makes it harder to do some things like resolve file paths to the proper sub-project.

Fixes https://github.com/applandinc/vscode-appland/issues/405

### CODE REVIEW

File: src/services/appmapUptodateService.ts [hunk 1]
```diff
@@ -93,6 +93,11 @@ export default class AppmapUptodateServiceInstance extends EventEmitter
     const result = new Set<string>();
     await Promise.all(
       [...outofDateAppMapFiles].map(async (file) => {
+        // Handle some anomalous behavior where the file path contains logging output:
+        if (file.startsWith('yarn run v')) return;
+        if (file.startsWith('Done in ')) return;
+        if (file.includes('appmap depends')) return;
```
Comment (dividedmind) [inline src/services/appmapUptodateService.ts:99]: I'm curious. Why and when would this be needed?

File: src/services/classMapIndex.ts [hunk 2]
```diff
@@ -268,125 +265,135 @@ export default class ClassMapIndex {
   protected async updateClassMap(): Promise<void> {
     if (this._dirty) {
       this._dirty = false;
-      this._codeObjectByFqid = {};
-      this._classMap = await this.buildClassMap();
+      const { codeObjectByFqid, classMap } = await backgroundJob<{
+        codeObjectByFqid: Record<string, CodeObjectEntry>;
+        classMap: CodeObjectEntry[];
+      }>('appmap.updateClassMap', () => buildClassMap(this.classMapFileURLs));
```
Comment (dividedmind) [inline src/services/classMapIndex.ts:271]: Won't the type parameter get automatically inferred from the `buildClassMap` return type?

**Review by dividedmind — COMMENTED** [review]
I haven't looked at how it improves performance with solidus yet, but the changes look good to me.

File: src/services/classMapIndex.ts [hunk 3]
```diff
@@ -4,24 +4,28 @@ import { readFile } from 'fs';
...
   FUNCTION = 'function',
 }
 
 export const InspectableTypes = [
-  CodeObjectEntryType.PACKAGE,
-  CodeObjectEntryType.CLASS,
-  CodeObjectEntryType.FUNCTION,
-  CodeObjectEntryType.QUERY,
-  CodeObjectEntryType.ROUTE,
+  CodeObjectEntryRootType.PACKAGE,
```
[the remark this answers is not in the document]
Comment (kgilpin) [reply src/services/classMapIndex.ts:25]: Thanks, I couldn't figure out the syntax for this.

[the remark this answers is not in the document]
Comment (kgilpin) [reply src/services/classMapIndex.ts:25]: I tried to make this change but I got a bunch of type errors that I didn't know how to resolve.

File: src/lib/backgroundJob.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+class Job<T> extends EventEmitter {
+  public timeout?: NodeJS.Timeout;
+
+  constructor(public fn: () => Promise<T>, public startTime: number, public previous?: Job<T>) {
+    super();
+  }
+
+  schedule(delayMs: number) {
+    const self = this as Job<T>;
+
```
[the remark this answers is not in the document]
Comment (kgilpin) [reply src/lib/backgroundJob.ts:14]: Without it, I get this warning:

Screen Shot 2022-06-19 at 1 04 57 PM

File: src/services/appmapUptodateService.ts (hunk 1 above)
Comment (kgilpin) [reply src/services/appmapUptodateService.ts:99]: I'm not sure, it's just something I observed and it cluttered up the console.

**Comment by kgilpin** [comment]
Thanks for the comments, I've fixed up what was quick and easy :-)

**Comment by dustinbyrne** [comment]
I haven't looked into it very deeply but I'm not getting any language detected (using `appmap-server`). We'll need to resolve this before we can ship it.

File: src/services/classMapIndex.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dividedmind) [reply src/services/classMapIndex.ts:25]: My bad, should have been
```typescript
export const CodeObjectEntryType = { ...CodeObjectEntryRootType, ...CodeObjectEntryChildType };
export type CodeObjectEntryType = CodeObjectEntryRootType | CodeObjectEntryChildType;
```

**Comment by kgilpin** [comment]
> I haven't looked into it very deeply but I'm not getting any language detected (using `appmap-server`). We'll need to resolve this before we can ship it.

This is a fix that I applied for this issue:

https://github.com/applandinc/vscode-appland/commit/2e8f930cbedf39dc0fec54e8c7a9fc0a5b385cc5#diff-d3a6792a8d41249e6294a65c5671e989a6a03034db5af3fe1f726fb465e42f08R33

**Review by dustinbyrne — APPROVED** [review]

**Pull request merged**